Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,591 @@
+
+            case .draining(let conn, _, let usedStreams):
+                self.state = .draining(conn, maxStreams: maxStreams, usedStreams: usedStreams)
+                return 0
+            }
+        }
+
+        mutating func goAwayReceived() -> EventLoop {
+            switch self.state {
+            case .starting, .backingOff, .closed:
+                preconditionFailure("Invalid state for draining a connection: \(self.state)")
+
+            case .active(let conn, let maxStreams, let usedStreams, _):
+                self.state = .draining(conn, maxStreams: maxStreams, usedStreams: usedStreams)
+                return conn.eventLoop
+
+            case .draining(let conn, _, _):
+                // we could potentially receive another go away while we drain all active streams and we just ignore it
+                return conn.eventLoop
+            }
+        }
+
+        /// The connection failed to start
+        mutating func failedToConnect() {
+            switch self.state {
+            case .starting:
+                self.state = .backingOff
+            case .backingOff, .active, .draining, .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+
+        mutating func fail() {
+            switch self.state {
+            case .starting, .active, .backingOff, .draining:
+                self.state = .closed
+            case .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+
+        mutating func lease(_ count: Int) -> Connection {
+            switch self.state {
+            case .starting, .backingOff, .draining, .closed:
+                preconditionFailure("Invalid state for leasing a stream: \(self.state)")
+
+            case .active(let conn, let maxStreams, var usedStreams, let lastIdle):
+                usedStreams += count
+                precondition(usedStreams <= maxStreams, "tried to lease a connection which is not available")
+                self.state = .active(conn, maxStreams: maxStreams, usedStreams: usedStreams, lastIdle: lastIdle)
+                return conn
+            }
+        }
+
+        /// should be called after a request has finished and the stream can be used again for a new request
+        /// - Returns: number of available streams which can be leased
+        mutating func release() -> Int {
+            switch self.state {
+            case .starting, .backingOff, .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+
+            case .active(let conn, let maxStreams, var usedStreams, var lastIdle):
+                usedStreams -= 1
+                assert(usedStreams >= 0)
+                if usedStreams == 0 {
+                    lastIdle = .now()
+                }
+                self.state = .active(conn, maxStreams: maxStreams, usedStreams: usedStreams, lastIdle: lastIdle)
+                return max(maxStreams - usedStreams, 0)
```

Comment:
I think `max` actually is necessary because we could receive a new max concurrent streams settings at any time which is lower than the currently streams in use. In that case, we would potentially report a negative number of available streams.

Nevertheless, I agree with raising the assert to a precondition and the unchecked subtraction.